---
title: "varroa virus networks"
author: "Sasha Mikheyev"
date: "1/22/2020"
output: html_document
---

```{r libraries}
library(tidyverse)
# awk '$3=="transcript"' GCF_002443255.1_Vdes_3.0_genomic.gff | perl -n -e 'print "$1\t$2\n" if /gene=LOC(\d+);.*transcript_id=(.*)/' | gzip >! gene2isoform.txt.gz
virusId <- read_tsv("data/viruses.txt", col_names = c("target_id", "description"))
```

```{r import}
read_kallisto <- function(filename) {
  sampleName <- sub("data/kallisto/(.*).tsv.gz","\\1", filename)
  return(read_tsv(filename) %>%
           select(!!sampleName := tpm))
}

df <- list.files(path = "data/kallisto", full.names = TRUE) %>% 
  lapply(read_kallisto) %>% 
  bind_cols() 

df$target_id <- list.files(path = "data/kallisto", full.names = TRUE)[1] %>% read_tsv() %>% select(target_id) %>% pull()

saveRDS(df, file = "data/kallisto.rds")
```

```{r}
left_join(virusId, df) %>% select(-target_id) %>% gather("library", "tpm", -description) %>% ggplot(aes(x= library, y = tpm, fill = description)) + geom_bar(stat = "identity") + theme(axis.text.x = element_text(angle = 90)) 

princomp(left_join(virusId, df) %>% select(-target_id) %>% gather("library", "tpm", -description) %>% spread(description, tpm) %>% select(-library)) %>% biplot(.)
```

